### ISSUE TITLE
got an error when using my own dataset

### ISSUE THREAD

**Issue by xutianq999** [PRBonn/point-cloud-prediction#9]
Hi Benedikt
Thank you so much for this work, I have achieved training and prediction with the help of your documentation
But I encountered difficulties when I wanted to use it on non-lidar point cloud data. This data comes from my depth camera [Bumblebeexb3]. The following is a sketch of it opened in Cloudcompare and in txt format.
![1672472465784](https://user-images.githubusercontent.com/82791384/210129345-27898351-5a87-4a9b-bba8-a7c05251fe50.png)
![754fd287d70b0d752352c6f9202ae97](https://user-images.githubusercontent.com/82791384/210129379-768b3f48-36fe-403f-9827-8c7f0cce4996.png)
The data contained in it is x y z intensity.It may not be obvious but it is a landslide of an indoor experiment, and I hope to predict the landslide through this work.
My problem is that he reported this error when I converted the txt file into a .bin file through the code.
![1672472611345](https://user-images.githubusercontent.com/82791384/210129449-cbd990d9-f8be-4c0a-812c-706185184fd7.png)
I use numpy to convert the data
`point_cloud = numpy.loadtxt("try.txt")
a.tofile("try.bin")`

When I use Cloudcompare to convert the txt file into .pcd and open it to add the intensity header file and then convert .pcd into .bin through another script, I can run through the above preprocessing link, but Dataloader does not read the data
![df9a5e2d5a1b38cfd280ab1580fed8c](https://user-images.githubusercontent.com/82791384/210129680-81bcd46f-4732-4fe8-9a00-1ecbd4da5d23.png)
![8371914b560bc8dd589584bdf37b43f](https://user-images.githubusercontent.com/82791384/210129684-75467e11-0fdb-43b2-867a-98b304ce3650.png)
You can also see that the calculated averages are all 0.
Could you please give me some suggestion or solution？

**Comment by benemer** [comment]
Hey @xutianq999!

The error you get during preprocessing is because we assume to have 3D points in the local frame of a rotating laser scanner with the field of view defined in `parameters.yaml`. During preprocessing, the point cloud is represented as a range image using a spherical projection.

You could try to process the disparity image of your stereo camera instead of the resulting 3D points. Given the baseline and the focal length of your setup, you can compute the depth values resulting in a 2D depth image. It might be easier to write your own [dataloader](https://github.com/PRBonn/point-cloud-prediction/blob/main/pcf/datasets/datasets.py#L221). Just make sure it returns past and future images and the required meta data. The images should be `torch.tensor` objects of size `(n_channels, n_steps, height, width)`, where the channel could be the disparity/depth in your case.

Also, you should disable the `CIRCULAR_PADDING` since this only holds for images with 360 degrees horizontal field of view.

Best
Benedikt

**Comment by xutianq999** [comment]
Thank you very much. I have been troubled by this problem for a long time and tried many ways. I did not expect you to reply so quickly. Once again, I would like to express my high respect to you.
Regarding the suggestion you mentioned, when I was reading the code, I noticed the part that forms the range_iamge by 3D point projection. I also tried to change some properties in the yaml file but did not achieve good results. I noticed that you also mentioned in your paper that CIRCULAR_PADDING will improve the data of the mechanic lidar to a certain extent.
Please forgive me and confirm to you again, I should rewrite Dataloader, not to read 3D point cloud files like .bin files, but to directly input depth maps, right? As the landslide data I provided above, can I convert it into a depth map, and then directly use such data for training?

Thank you again for your reply, you have given me great encouragement and courage

**Comment by benemer** [comment]
You are welcome! Yes, the circular pattern helps to propagate features from one side of the range image to the other one. This is important since the rotating LiDAR covers the whole 360 degrees in the azimuthal direction, which means that the left and right range image boundaries are actually connected. Since this does not hold for your depth image, you should disable the circular padding.

Regarding your problem: Yes, instead of reading and projecting 3D point clouds, you can directly load the depth maps in the dataloader. Just modify the [`__getitem__`](https://github.com/PRBonn/point-cloud-prediction/blob/main/pcf/datasets/datasets.py#L178) as mentioned above. Also, the `MIN_RANGE` and `MAX_RANGE` in the `parameters.yml` are used as limits for the prediction. If you want to predict the depth, these should be the minimum and maximum depth you want to predict. You also need to change some parameters like the `HEIGHT` and `WIDTH` of your depth maps.

For now, I would train the model without using the Chamfer Distance loss and only use the range/depth L1 loss and the valid mask loss. If you want to use Chamfer Distance loss as well, you need to modify the `get_target_mask_from_range_view` method to get the 3D point cloud from your predicted depth map.

So, there are some modifications needed to make this work with depth maps, but it should be possible. I actually used this codebase and modified it for the [Waymo Occupancy and Flow Prediction Challenge](https://waymo.com/open/challenges/2022/occupancy-flow-prediction-challenge/) which worked quite well.

Best regards
Benedikt

**Comment by yzysmile** [comment]
Hi Benedikt, 
I watched your communication process. I have the following thoughts, I don’t know if they are correct or not. 

1. If i input depth maps directly for traing, there is no need the process of projection and i do not need to calculate the depth value by the baseline and focal length because each depth map itself stores the depth value.
2. Each ground truth mask has the same size as the depth map and the value of each pixel position is 1 because each pixel of the depth map corresponds to a three-dimensional point.
3. Depth images can be convert to cloud points by the camera intrisinc, so I still calculate Chamfer Distance loss easily.

Looking forward to your reply.

best wishes
yzy

**Comment by benemer** [comment]
Hey,

1. Yes, indeed. You just have to modify the dataloader to read your depth maps. I did not try to train it on depth maps, but I don't see a reason why it should not work.
2. If you have pixels without/with invalid depth information, the mask should be zero, and therefore ignore these. This happens for example for the sky.
3. If you compute the 3D point, this should be possible. Again, you have to modify the code for this.

Best
Benedikt

**Comment by yzysmile** [comment]
> Hey,
> 
> 1. Yes, indeed. You just have to modify the dataloader to read your depth maps. I did not try to train it on depth maps, but I don't see a reason why it should not work.
> 2. If you have pixels without/with invalid depth information, the mask should be zero, and therefore ignore these. This happens for example for the sky.
> 3. If you compute the 3D point, this should be possible. Again, you have to modify the code for this.
> 
> Best Benedikt

Thank you for your quick replies one by one.
 I will try to modify the code and wish you a Merry Christmas and good scientific research!

Best 
yzy
